File: Examples/ObjC/Headless/SlyceSDK-Headless-Demo-ObjC/Camera.m

#import "Camera.h"

@interface Camera () <AVCapturePhotoCaptureDelegate>

@property (nonatomic, strong) AVCaptureVideoPreviewLayer *cameraPreviewLayer;

@end

@implementation Camera

- (AVCaptureVideoPreviewLayer *)start {
    if (nil == self.cameraCaptureSession) {
        self.cameraCaptureSession = [self setupCaptureSession];
        
        if (!self.cameraCaptureSession) {
            NSLog(@"Unable to create a camera capture session; this means the camera is unavailable or impossible to access and none of the SDK features will work!");
            return nil;
        }
    }
    
    if (nil == self.cameraPreviewLayer) {
        AVCaptureVideoPreviewLayer *previewLayer = [[AVCaptureVideoPreviewLayer alloc] initWithSession:self.cameraCaptureSession];
        previewLayer.videoGravity = AVLayerVideoGravityResizeAspectFill;
        self.cameraPreviewLayer = previewLayer;
        
        if (nil == self.cameraPreviewLayer) {
            NSLog(@"Unable to create a camera preview layer; this means the camera will be invisible in the demo UI and it is unlikely that the SDK features will work!");
            return nil;
        }
    }
    
    
    if (self.cameraCaptureSession.isRunning == false) {
        [self.cameraCaptureSession startRunning];
    }
    
    return self.cameraPreviewLayer;
}


- (AVCaptureDevice *_Nullable)getAvailableCamera {
    return [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeVideo];
}

- (void)prepareCamera:(AVCaptureDevice *)camera {
    NSError *error;
    [camera lockForConfiguration:&error];
    
    if (error != nil) {
        NSLog(@"%@", error.localizedDescription);
        return;
    }
    
    camera.focusMode = AVCaptureFocusModeContinuousAutoFocus;
    [camera unlockForConfiguration];
}

- (AVCaptureSession * _Nullable)setupCaptureSession {
    // create session
    AVCaptureSession *captureSession = [[AVCaptureSession alloc] init];
    captureSession.sessionPreset = AVCaptureSessionPresetHigh;
    
    // create, setup camera
    AVCaptureDevice *camera = [self getAvailableCamera];
    if (!camera) {
        return nil;
    }
    
    [self prepareCamera:camera];
    
    
    // add camera input to session
    NSError *error;
    AVCaptureDeviceInput *cameraInput = [[AVCaptureDeviceInput alloc] initWithDevice:camera error:&error];
    if (error != nil) {
        NSLog(@"%@", error.localizedDescription);
        return nil;
    }
    
    if (![captureSession canAddInput:cameraInput]) {
        return nil;
    }
    
    [captureSession addInput:cameraInput];
    return captureSession;
}

@end

File: Examples/Swift/Headless/SlyceSDK-Headless-Demo-Swift/Camera.swift
import UIKit
import AVFoundation

class Camera: NSObject, AVCapturePhotoCaptureDelegate {
    public var cameraCaptureSession: AVCaptureSession?
    var cameraPreviewLayer: AVCaptureVideoPreviewLayer?
    
    func start() -> AVCaptureVideoPreviewLayer? {
        if nil == cameraCaptureSession {
            cameraCaptureSession = setupCaptureSession()
            
            if nil == cameraCaptureSession {
                print("Unable to create a camera capture session; this means the camera is unavailable or impossible to access and none of the SDK features will work!")
                return nil
            }
        }
        
        if nil == cameraPreviewLayer {
            let previewLayer = AVCaptureVideoPreviewLayer(session: cameraCaptureSession!)
            previewLayer.videoGravity = AVLayerVideoGravity.resizeAspectFill
            cameraPreviewLayer = previewLayer
            
            if nil == cameraPreviewLayer {
                print("Unable to create a camera preview layer; this means the camera will be invisible in the demo UI and it is unlikely that the SDK features will work!")
                return nil
            }
        }
        
        if cameraCaptureSession!.isRunning == false {
            cameraCaptureSession?.startRunning()
        }
        
        return cameraPreviewLayer
    }
    
    func getAvailableCamera() -> AVCaptureDevice? {
        return AVCaptureDevice.default(for: .video)
    }
    
    func prepareCamera(_ camera: AVCaptureDevice) {
        do {
            try camera.lockForConfiguration()
            camera.focusMode = .continuousAutoFocus
            camera.unlockForConfiguration()
        } catch let e {
            print(e)
        }
    }
    
    func setupCaptureSession() -> AVCaptureSession? {
        // create session
        let captureSession = AVCaptureSession()
        captureSession.sessionPreset = .high
        
        // create, setup camera
        guard let camera = getAvailableCamera() else {
            return nil
        }
        
        self.prepareCamera(camera)
        
        // add camera input to session
        var cameraInput: AVCaptureDeviceInput
        do {
            try cameraInput = AVCaptureDeviceInput(device: camera)
        } catch let e {
            print(e)
            return nil
        }
        
        if false == captureSession.canAddInput(cameraInput) {
            return nil
        }
        
        captureSession.addInput(cameraInput)
        return captureSession
    }
}
